documentation/src/docs/watermarking.md

```diff
@@ -0,0 +1,438 @@
+
+## Config API
+The `StreamConfiguration` would have the new parameter:
+
+```
+/**
+ * The duration after the last call to {@link EventStreamWriter#noteTime(long)} which the
+ * timestamp should be considered valid before it is forgotten. Meaning that after this long of
+ * not calling {@link EventStreamWriter#noteTime(long)} the writer will be forgotten.
+ * If there are no known writers, readers that call {@link EventStreamReader#getCurrentTimeWindow()}
+ * will receive a `null` when they are at the corresponding position in the stream.
+ */
+private final long timestampAggregationTimeout;
+```
+The `EventWriterConfig` would have the new parameter:
+```
+/**
+     * Automatically invoke {@link EventStreamWriter#noteTime(long)} passing
+     * {@link System#currentTimeMillis()} on a regular interval.
+     */
+    private final boolean automaticallyNoteTime;
+
+```
+Because it may take some time for all the Writers to come online, for a new Stream we may also want to add an "intentional delay" before writing the first mark to the marks Segment. This may or may not be configured on a per-stream basis.
+
+## Event Reader API
+On the Reader side, a new method is added on the `EventStreamReader` to obtain the location of the Reader in terms of time:
+
+```
+/**
+     * Returns a window which represents the range of time that this reader is currently reading as
+     * provided by writers via the {@link EventStreamWriter#noteTime(long)} API.
+     *
+     * If no writers were providing timestamps at the current position in the stream `null` will be returned.
+     *  
+     * @return A TimeWindow which bounds the current location in the stream, or null if one cannot be established.
+     */
+    TimeWindow getCurrentTimeWindow();
+
+```
+Where `TimeWindow` is defined as:
+
+```
+/**
+ * Represents a time window for the events which are currently being read by a reader.
+ *
+ * The upper time bound is a timestamp which is greater than or equal to any that were provided by
+ * any writer via the {@link EventStreamWriter#noteTime(long)} API prior to the current location in
+ * the stream.
+ *
+ * Similarly the lower time bound is a timestamp which is less than or equal to the most recent
+ * value provided via the {@link EventStreamWriter#noteTime(long)} API for by any writer using that
+ * API at the current location in the stream.
+ *
+ * upperTimeBound will always be greater than or equal to lowerTimeBound.
+ */
+@Data
+public class TimeWindow {
+    private final long lowerTimeBound;
+    private final long upperTimeBound;
+}
+
+```
+
+Note that the `TimeWindow` that is seen by Readers can be affected by `timestampAggrigationTimeout` in the config above. Different applications may wish to be more or less conservative as to when Writers are 'alive' and need to be considered.
+
+# Internal API Changes
```

Comment:
Changes?

Also I'm not clear on why internal APIs need to be part of our public docs.

Reply:
@tkaitchuck removed